integration-tests/chaos/congestion/simulator.go

```diff
@@ -0,0 +1,484 @@
+			for range txsToSend {
+				select {
+				case work <- struct{}{}:
+				case <-ctx.Done():
+					return nil
+				}
+			}
+			return nil
+		})
+	}()
+}
+
+func (s *Simulator) runPhaseController(ctx context.Context, wg *sync.WaitGroup) {
+	defer wg.Done()
+	s.lggr.Info("Starting ramp up phase")
+	s.phase.Store(phaseRampUp)
+	s.updateCurrentFees(0)
+	s.handleVolatilePhase(ctx, s.RampUp, true)
+
+	s.lggr.Info("Starting plateau phase")
+	s.phase.Store(phasePlateau)
+	s.expectedCurrentCongestion.Store(s.Plateau.Congestion)
+	timer := time.NewTimer(time.Second * time.Duration(s.Plateau.Duration))
+	defer timer.Stop()
+	select {
+	case <-timer.C:
+		break
+	case <-ctx.Done():
+		return
+	}
+
+	s.lggr.Info("Starting cooldown phase")
+	s.phase.Store(phaseCoolDown)
+	s.handleVolatilePhase(ctx, s.CoolDown, false)
+
+	s.lggr.Info("Finished iteration")
+	s.phase.Store(phaseInactive)
+}
+
+func (s *Simulator) handleVolatilePhase(ctx context.Context, phase Phase, isIncrease bool) {
+	ticker := time.NewTicker(time.Second)
+	defer ticker.Stop()
+	phaseTimer := time.NewTimer(time.Second * time.Duration(phase.Duration))
+	defer phaseTimer.Stop()
+	phaseStart := time.Now()
+	s.expectedCurrentCongestion.Store(phase.Congestion)
+phaseLoop:
+	for {
+		select {
+		case <-ticker.C:
+		case <-phaseTimer.C:
+			break phaseLoop
+		case <-ctx.Done():
+			return
+		}
+
+		elapsedRatio := min(float64(time.Since(phaseStart))/float64(time.Duration(phase.Duration)*time.Second), 1)
+		if !isIncrease {
+			// count backwards. If 0.1 of cooldown phase has passed, we should be at the same price point as if 0.9 of ramp up phase has passed
+			elapsedRatio = 1 - elapsedRatio
+		}
+		s.updateCurrentFees(elapsedRatio)
+	}
+
+	if isIncrease {
+		s.updateCurrentFees(1)
+	} else {
+		s.updateCurrentFees(0)
+	}
+}
+
+// updateCurrentFees - sets fees according to expected distance from target. If targetDistanceRatio = 1, current fees s.FeesIncreasePercent compared to initial.
+func (s *Simulator) updateCurrentFees(targetDistanceRatio float64) {
+	targetPercent := s.FeesIncreasePercent * targetDistanceRatio
+	baseFee := uint64(float64(s.InitialBaseFeePerGas) * (1 + targetPercent/100))
+	s.expectedCurrentBaseFee.Store(baseFee)
+	tip := uint64(float64(s.InitialTipPerGas) * (1 + targetPercent/100))
+	s.expectedCurrentTip.Store(tip)
+	s.lggr.Debugf("Updating fees to %d according to distance %.2f", baseFee, targetDistanceRatio)
+}
+
+func (s *Simulator) runTick(ctx context.Context) {
+	var wg sync.WaitGroup
+	ctx, cancel := context.WithTimeout(ctx, time.Second*time.Duration(s.RampUp.Duration+s.Plateau.Duration+s.CoolDown.Duration))
+	defer cancel()
+	wg.Add(1)
+	go s.runPhaseController(ctx, &wg)
+	s.startTransactionProducer(ctx, &wg)
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		s.listenHeadsUntilCanceled(ctx, s.updateBaseFee)
+	}()
+	wg.Wait()
+}
+
+func (s *Simulator) updateBaseFee(_ context.Context, _ *types.Header) error {
```

Comment:
any reason for why this function isn't just `updateBaseFee() error`?

Reply:
It's called on every block in `listenHeads`, that's why accepts arguments that are not used.